src/overlays/actors/ovl_En_Knight/z_en_knight.c

```diff
@@ -205,237 +380,4044 @@ ActorInit En_Knight_InitVars = {
+    this->actionFunc = EnKnight_GetBackUp;
+    this->actionTimers[0] = 35;
+}
+
+void EnKnight_GetBackUp(EnKnight* this, PlayState* play) {
+    SkelAnime_Update(&this->skelAnime);
+    this->neckPitchTarget = 0;
+    Math_ApproachF(&this->shadowAlphaFactor, 180.0f, 1.0f, 10.0f);
+
+    if (this->actionTimers[0] == 0) {
+        EnKnight_SetupWait(this, play);
+    }
+
+    if ((this->actionTimers[0] > (s16)(BREG(74) + 10)) && !(this->randTimer & (BREG(76) + 3))) {
+        EnKnight* afterImage = (EnKnight*)Actor_SpawnAsChild(
+            &play->actorCtx, &this->actor, play, ACTOR_EN_KNIGHT, this->actor.world.pos.x, this->actor.world.pos.y,
+            this->actor.world.pos.z, this->actor.world.rot.x, this->actor.world.rot.y, this->actor.world.rot.z,
+            (this == sIgosInstance) ? EN_KNIGHT_PARAM_IGOS_BODY_AFTERIMAGE : EN_KNIGHT_PARAM_KNIGHT_BODY_AFTERIMAGE);
+
+        if (afterImage != NULL) {
+            s32 i;
+
+            for (i = 0; i < KNIGHT_LIMB_MAX; i++) {
+                afterImage->skelAnime.jointTable[i] = this->skelAnime.jointTable[i];
+                afterImage->skelAnime.skeleton = this->skelAnime.skeleton;
+                afterImage->skelAnime.dListCount = this->skelAnime.dListCount;
+            }
+        }
+    }
+}
+
+void EnKnight_SetupStunned(EnKnight* this, PlayState* play, u16 duration) {
+    this->actionFunc = EnKnight_Stunned;
+    this->actionTimers[0] = duration;
+    this->actionTimers[1] = 30;
+    Actor_SetColorFilter(&this->actor, COLORFILTER_COLORFLAG_BLUE, 120, 0, duration);
+}
+
+void EnKnight_Stunned(EnKnight* this, PlayState* play) {
+    if (this->actionTimers[1] > 0 && this->actionTimers[1] < 23) {
+        if (this->actionTimers[1] == 1) {
+            sTargetKnight = NULL;
+        } else {
+            sTargetKnight = this;
+        }
+    }
+
+    Math_ApproachZeroF(&this->actor.speed, 1.0f, 1.0f);
+
+    if (this->actionTimers[0] < 20) {
+        if (this->actionTimers[0] % 2 != 0) {
+            this->neckPitch = 0x708;
+        } else {
+            this->neckPitch = -0x708;
+        }
+    } else {
+        this->neckPitchTarget = 0;
+    }
+
+    if (this->actionTimers[0] == 0) {
+        EnKnight_SetupWait(this, play);
+    }
+}
+
+void EnKnight_SetupRecoilFromDamage(EnKnight* this, PlayState* play) {
+    Vec3f translation;
+    s16 yawDiff;
+
+    this->actionFunc = EnKnight_RecoilFromDamage;
+
+    yawDiff = this->yawToPlayer - this->actor.shape.rot.y;
+    if (ABS_ALT(yawDiff) < 0x4000) {
+        Animation_MorphToPlayOnce(&this->skelAnime, &gKnightDamagedFrontAnim, 0.0f);
+        this->subAction = KNIGHT_SUB_ACTION_RECOIL_0;
+    } else {
+        Animation_MorphToPlayOnce(&this->skelAnime, &gKnightDamagedBackAnim, 0.0f);
+        this->subAction = KNIGHT_SUB_ACTION_RECOIL_1;
+    }
+
+    Matrix_RotateYS(this->yawToPlayer, MTXMODE_NEW);
+    Matrix_MultVecZ(-15.0f, &translation);
+    this->animTranslationX = translation.x;
+    this->animTranslationZ = translation.z;
```

Comment:
Morita bosses love doing this, and I'm not sure on a good name for it. For Odolwa, I used `additionalVelocityX/Z`, since it was used for *both* some additional translation during certain animations. But for Bio Baba, I just called it `knockbackVelocity` (it's a Vec3f in that case) because it was only used for getting knocked back from damage. Majora has `knockbackVelocityX/Z` in a similar fashion. I used "velocity" in all these cases because the units are implicitly units/frame.

The use case in this actor *seems* to be similar to Odolwa, but I've never been the biggest fan of that name in the first place. I'm not sure if you have some grand unifying idea of how these should all be named with that extra context; it would be nice to get these at least somewhat consistent since they appear semi-frequently. I'm curious how you came up with `animTranslationX/Z` and what your thoughts are on the other names?

Reply:
I like translation here since it's a position offset. To me it seems like this variable is a general purpose position offset used for animations that would ideally have some translation component, but the animation system doesn't treat animations with translations in their data particularly nicely so Morita did the classic "I'll just implement this myself".

I dislike velocity since it's not in the same units as typical velocities (since they're scaled by the framerate divisor), it feels clearer to describe it as a translation/offset/displacement

Reply:
I didn't think about how this might have come about to add translations to animations; that's an interesting perspective I hadn't considered. However, I took a look at the way other similar actors use this and I'm not 100% convinced this is the case, at least universally. An equivalent variable in Boss05 is *only* used for these little bumps off my shield, which don't actually interrupt or change its animation at all; it just keeps doing what it was doing before but with a little push:
https://github.com/user-attachments/assets/a8afc14d-b1c0-4694-b61e-9f0030977605

But this might go back to what I said in my original comment; maybe there *is* no unifying name for all of these. Perhaps it's animation-related for Odolwa and Knight and bump/knockback-related for Bio Babas, and they should have different names. But I figured I'd bring it up anyways to see if that changes your perspective on things. I'm not saying it *has* to change your perspective or anything; I'm just curious.

As for translation vs. velocity, maybe I'm just dumb, but I don't see the way that traditional velocities get scaled by the framerate divisor? I checked every usage of `gFramerateDivisor` in the codebase and I didn't see anything relevant. I did check every usage of these kind of variables in MM3D (which is why this reply took so long; sorry about that!) and they just multiplied them by (2.0f / 3) before applying them to the world position, e.g.:
```c
this->actor.world.pos.x += this->animTranslationX * (2.0f / 3);
this->actor.world.pos.z += this->animTranslationZ * (2.0f / 3);
```
This makes sense, because otherwise the translation would be too large given the game's higher framerate. However, they *also* do this same sort of calculation whenever they set things like `this->actor.speed`, so I sort of took that to mean they were at least somewhat similar.

Reply:
I guess it actually does make sense to think of them like velocities in how they behave, since it gets applied every frame. I was thinking of a parallel between this variable and `CollisionCheckInfo::displacement` at one point, but this field is reset every frame (cf. `CollisionCheck_ResetDamage`) so it isn't really comparable. It's pretty unfortunate though that they're in a different scale to the "usual" velocities, it seems confusing to refer to both without any distinction.

It seems like in general this variable covers the need to move the actor in some controlled direction (independent of the world rotation that move functions like `Actor_MoveWithGravity` use to decompose the speed variable) and not have it reset every frame, for various purposes (like knockback or for certain animations as we saw). If we'd really like to cover all cases I can't think of a name that isn't very generic, like `additionalVelocity` or `additionalMovement`.

Reply:
Yeah, I think I agree with all of that. I didn't think of comparing it to `CollisionCheckInfo`'s `displacement` myself, but like you noted, it's not really the same thing. And yeah, it's definitely weird that it's not scaled like a "normal" velocity too.

It's unfortunate that there really is no good term for this that (a) encompasses all use cases across all actors, and (b) isn't super generic, so I'm fine with this being unique from other Morita actors. Your comment did make me think about one alternative here though; what do you think about `animMovementX` and `animMovementZ`? If you don't like it, that's fine; I'm okay with these being `animTranslationX` and `animTranslationZ` now. But I figured I'd bring it up, since you brought up `additionalMovement`. My thinking is it sort of straddles the line where it's clear that it isn't *exactly* like a "usual" velocity, but it's close. Though maybe it's more confusing than anything else.

Thanks for your patience and detail in discussing this. Seeing someone else's viewpoint here really helped shape my own thoughts on these kinds of variables, so I really appreciate it.

Reply:
I prefer `animMovement` over `animTranslation` at least. Changed it